[Automation] Add CI linters/checks (#190)
pokt-network/pocket: docs/development/README.md

├── utility          # Implementation of the Utility module
├── Makefile         # [to-be-deleted] The source of targets used to develop, build and test
```

> ### Linters

Reviewer: Please add a section explaining "how does one add new linters?"

Author: @Olshansk I've added more documentation, including the part on how to manage linters, and how to write in-house checks/rules.

Reviewer: LGTM!

Our CI automation runs some linters and static checks to ensure code quality. You can run them locally with:

```bash
make lint
```

> You might need to install some of the linters locally, here is a list of them.

Reviewer: Where possible/simple, can you add a make target that runs the commands to install them and update the docs here?

Author: @Olshansk I think it's better to keep the link to installation instructions. Snippet from `golangci-lint` docs:
> such `go install`/`go get` installation aren't guaranteed to work. We recommend using binary installation.

I'm not aware of other OS-agnostic way of installing the binary we could use in Makefile. :(

Reviewer: Cool. Thanks for looking into it!

├── utility          # Implementation of the Utility module
├── Makefile         # [to-be-deleted] The source of targets used to develop, build and test
```

### Linters

> We are utilizing `golangci-lint` to run the linters. It is a wrapper around a number of linters and is configured to run many of them at once. The linters are configured to run on every commit and pull request via CI, and all code issues are populating as GitHub annotations to let developers and reviewers easily locate an issue.

Reviewer: ```suggestion
We utilize `golangci-lint` to run the linters. It is a wrapper around a number of linters and is configured to run many at once. The linters are configured to run on every commit and pull request via CI, and all code issues are populated as GitHub annotations to let developers and reviewers easily locate an issue.
```

#### Installation of golangci-lint

> Please follow the instructions on the [golangci-lint](https://golangci-lint.run/usage/install/#local-installation) website.

Reviewer: Make target that helps with installation?

#### Running linters locally

You can run `golangci-lint` locally against all packages with:

```bash
make lint
```

If you need to specify any additional flags, you can run `golangci-lint` directly as it picks up the configuration from the `.golangci.yml` file.

#### VSCode Integration

> `golangci-lint` has an integration with VSCode. Per [documentation](https://golangci-lint.run/usage/integrations/), recommended settings are:

Reviewer: <img width="312" alt="Screen Shot 2022-09-27 at 6 02 18 PM" src="https://user-images.githubusercontent.com/1892194/192663994-a0d609c5-3ea4-4d00-b1de-d6f91869a85f.png">

Configured.

```json
"go.lintTool":"golangci-lint",
"go.lintFlags": [
  "--fast"
]
```

#### Configuration

> `golangci-lint` is a sophisticated tool as it includes a lot of linters, including our own custom linting rules. As a result, a configuration file might become quite large. The configuration file is located at [`.golangci.yml`](../../.golangci.yml).

Reviewer: ```suggestion
`golangci-lint` is a sophisticated tool including both default and custom linters. The configuration file, which can grow quite large, is located at [`.golangci.yml`](../../.golangci.yml).
```

The official documentation includes a list of different linters and their configuration options. Please refer to [this page](https://golangci-lint.run/usage/linters/) for more information.

#### Custom linters

> We can write custom linters using [`go-ruleguard`](https://go-ruleguard.github.io/). The rules are located in the [`misc/linting-rules`](../../misc/linting-rules) directory. The rules are written in the [Ruleguard DSL](https://github.com/quasilyte/go-ruleguard/blob/master/_docs/dsl.md), if you've never worked with ruleguard in the past, it makes sense to go through [introduction article](https://quasilyte.dev/blog/post/ruleguard/) and [Ruleguard by example tour](https://go-ruleguard.github.io/by-example/).

Reviewer: cc @kutoft regarding this entire README.

Once it is merged to main, I will ping the team directly in discord, but there's no reason for your team not to leverage all this infra.
